File: zhengshan-hmi/ConfigToolNew/HmiConfig/Form/HmiData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Collections;

namespace HmiConfig
{
    class xValveCtrlCmd
    { 
	    public UInt16 dBOffsetForOpenCmd;//offset in the database
        public UInt16 openCmd;
        public UInt16 dBOffsetForCloseCmd;//offset in the databse
        public UInt16 closeCmd;
        public UInt16 dBoffsetForStopCmd;//offset in the databse
        public UInt16 stopCmd;
        public UInt16 dBOffsetForCurPosCmd;//offset in the databse
        public UInt16 getCurPosCmd;
        public UInt16 dbOffsetForCurPos;//offset in the database
        public UInt16 curPosMaxVal;

        public xValveCtrlCmd()
        {
            dBOffsetForOpenCmd = 0xFFFF;
            openCmd = 0xFFFF;
            dBOffsetForCloseCmd = 0xFFFF;//offset in the databse
            closeCmd = 0xFFFF;
            dBoffsetForStopCmd = 0xFFFF;//offset in the databse
            stopCmd = 0xFFFF;
            dBOffsetForCurPosCmd = 0xFFFF;//offset in the databse
            getCurPosCmd = 0xFFFF;
            dbOffsetForCurPos = 0xFFFF;//offset in the database
            curPosMaxVal = 100;
        }

        public xValveCtrlCmd(byte[] b)
        {
            int offset = 0;
            dBOffsetForOpenCmd = (UInt16)(b[offset++] | (b[offset++] << 8));
            openCmd = (UInt16)(b[offset++] | (b[offset++] << 8));
            dBOffsetForCloseCmd = (UInt16)(b[offset++] | (b[offset++] << 8));//offset in the databse
            closeCmd = (UInt16)(b[offset++] | (b[offset++] << 8));
            dBoffsetForStopCmd = (UInt16)(b[offset++] | (b[offset++] << 8));//offset in the databse
            stopCmd = (UInt16)(b[offset++] | (b[offset++] << 8));
            dBOffsetForCurPosCmd = (UInt16)(b[offset++] | (b[offset++] << 8));//offset in the databse
            getCurPosCmd = (UInt16)(b[offset++] | (b[offset++] << 8));
            dbOffsetForCurPos = (UInt16)(b[offset++] | (b[offset++] << 8));//offset in the database
            curPosMaxVal = (UInt16)(b[offset++] | (b[offset++] << 8));
        }
        
        public byte[] GetData()
        {
            byte[] b= new byte[DataSizeof];
            int offset = 0;
            b[offset++] = (byte)dBOffsetForOpenCmd;
            b[offset++] = (byte)(dBOffsetForOpenCmd>>8);

            b[offset++] = (byte)openCmd;
            b[offset++] = (byte)(openCmd>>8);

            b[offset++] = (byte)dBOffsetForCloseCmd;
            b[offset++] = (byte)(dBOffsetForCloseCmd>>8);

            b[offset++] = (byte)closeCmd;
            b[offset++] = (byte)(closeCmd>>8);

            b[offset++] = (byte)dBoffsetForStopCmd;
            b[offset++] = (byte)(dBoffsetForStopCmd>>8);

            b[offset++] = (byte)stopCmd;
            b[offset++] = (byte)(stopCmd>>8);

            b[offset++] = (byte)dBOffsetForCurPosCmd;
            b[offset++] = (byte)(dBOffsetForCurPosCmd>>8);

            b[offset++] = (byte)getCurPosCmd;
            b[offset++] = (byte)(getCurPosCmd>>8);

            b[offset++] = (byte)dbOffsetForCurPos;
            b[offset++] = (byte)(dbOffsetForCurPos>>8);

            b[offset++] = (byte)curPosMaxVal;
            b[offset++] = (byte)(curPosMaxVal >> 8);

            return b;
        }

        public static int DataSizeof
        {
            get
            {
                return sizeof(UInt16)*10;
            }
        }
    }
    class xValveView
    {
        public UInt32 dBOffsetForOpenVal;
        public UInt32 dBOffsetForCloseVal;
        public UInt32 dBOffsetForLRVal;
        public UInt32 dBOffsetForCurPos;
        public UInt32 curPosMaxVal;
        public xValveView()
        {
            dBOffsetForOpenVal = 0xFFFFFFFF;
            dBOffsetForCloseVal = 0xFFFFFFFF;
            dBOffsetForLRVal = 0xFFFFFFFF;
            dBOffsetForCurPos = 0xFFFFFFFF;
            curPosMaxVal = 100;     
        }
        public xValveView(byte[] b)
        {
            int offset = 0;
            dBOffsetForOpenVal = (UInt32)(b[offset++] | (b[offset++] << 8) | (b[offset++] << 16) | (b[offset++] << 24));
            dBOffsetForCloseVal = (UInt32)(b[offset++] | (b[offset++] << 8) | (b[offset++] << 16) | (b[offset++] << 24)); 
            dBOffsetForLRVal = (UInt32)(b[offset++] | (b[offset++] << 8) | (b[offset++] << 16) | (b[offset++] << 24));
            dBOffsetForCurPos = (UInt32)(b[offset++] | (b[offset++] << 8) | (b[offset++] << 16) | (b[offset++] << 24));
            curPosMaxVal = (UInt32)(b[offset++] | (b[offset++] << 8) | (b[offset++] << 16) | (b[offset++] << 24)); 
        }
        public byte[] GetData()
        {
            byte[] b= new byte[DataSizeof];
            int offset = 0;
            b[offset++] = (byte)dBOffsetForOpenVal;
            b[offset++] = (byte)(dBOffsetForOpenVal >> 8);
            b[offset++] = (byte)(dBOffsetForOpenVal>>16);
            b[offset++] = (byte)(dBOffsetForOpenVal>>24);

            b[offset++] = (byte)dBOffsetForCloseVal;
            b[offset++] = (byte)(dBOffsetForCloseVal>>8);
            b[offset++] = (byte)(dBOffsetForCloseVal>>16);
            b[offset++] = (byte)(dBOffsetForCloseVal>>24);

            b[offset++] = (byte)dBOffsetForLRVal;
            b[offset++] = (byte)(dBOffsetForLRVal>>8);
            b[offset++] = (byte)(dBOffsetForLRVal>>16);
            b[offset++] = (byte)(dBOffsetForLRVal>>24);

            b[offset++] = (byte)dBOffsetForCurPos;
            b[offset++] = (byte)(dBOffsetForCurPos>>8);
            b[offset++] = (byte)(dBOffsetForCurPos>>16);
            b[offset++] = (byte)(dBOffsetForCurPos>>24);

            b[offset++] = (byte)curPosMaxVal;
            b[offset++] = (byte)(curPosMaxVal>>8);
            b[offset++] = (byte)(curPosMaxVal>>16);
            b[offset++] = (byte)(curPosMaxVal>>24);

            return b;
        }
        public static int DataSizeof
        {
            get
            {
                return sizeof(UInt32) * 5;
            }
        }
    }

    class xValveData
    { 
        public const int VALVE_NAME_MAX_LENTH = 15;
    	public byte[] acValveName = new byte[VALVE_NAME_MAX_LENTH+1];
	    public xValveCtrlCmd xCmd;// = new xValveCtrlCmd();
	    public xValveView xView;// = new xValveView();

        public xValveData()
        {
            for (int i = 0; i < acValveName.Length; i++)
            {
                acValveName[i] = 0;
            }
            xCmd = new xValveCtrlCmd();
	        xView = new xValveView();
        }

        public xValveData(byte[] b)
        {

            int offset = 0;
            Array.Copy(b, offset, acValveName, 0, acValveName.Length);
            offset += (((acValveName.Length + 1) >> 1) << 1);
            byte[] xValveCtrlCmdData = new byte[xValveCtrlCmd.DataSizeof];
            Array.Copy(b, offset, xValveCtrlCmdData, 0, xValveCtrlCmd.DataSizeof);
            offset = (((xValveCtrlCmd.DataSizeof + acValveName.Length  + 3) >> 2) << 2);
            xCmd = new xValveCtrlCmd(xValveCtrlCmdData);
            byte[] xValveViewData = new byte[xValveView.DataSizeof];
            Array.Copy(b, offset, xValveViewData, 0, xValveView.DataSizeof);
            //offset += xValveView.DataSizeof;
            xView = new xValveView(xValveViewData);

        }
        public byte[] GetData()
        {
            int offset=0;
            byte[] b= new byte[DataSizeof];
            
            Array.Copy(acValveName, 0, b, 0, acValveName.Length);
            
            offset += (((acValveName.Length + 1) >> 1) << 1);
            Array.Copy(xCmd.GetData(), 0, b, offset, xValveCtrlCmd.DataSizeof);
            offset = (((xValveCtrlCmd.DataSizeof + acValveName.Length  + 3) >> 2) << 2);
            Array.Copy(xView.GetData(), 0, b, offset, xValveView.DataSizeof);
            return b;
        }
        
        public static int DataSizeof
        {
            get
            {
                return xValveCtrlCmd.DataSizeof + xValveView.DataSizeof + (((VALVE_NAME_MAX_LENTH + 1 + 3) >> 2) <<2);
            }
        }
    }
    class xValveGroup 
    {
        public const int GROUP_VALVE_NUM = 40; 
        public xValveData[] xValve = new xValveData[GROUP_VALVE_NUM];
        public xValveGroup()
        {
            for (int i = 0; i < GROUP_VALVE_NUM; i++)
            {
                xValve[i] = new xValveData();
            }
        }
        public xValveGroup(byte[] b)
        {
            int offset = 0;
            byte[] xValveDataData = new byte[xValveData.DataSizeof];
            for (int i = 0; i < GROUP_VALVE_NUM; i++)
            {
                Array.Copy(b, offset, xValveDataData, 0, xValveData.DataSizeof);
                xValve[i] = new xValveData(xValveDataData);
                offset += xValveData.DataSizeof;
            }
        }
        
        public byte[] GetData()
        {
            ArrayList arrTemp = new ArrayList(); 
            for (int i = 0; i < GROUP_VALVE_NUM; i++)
            {
                arrTemp.AddRange(xValve[i].GetData()); 
            }

            return (byte[])arrTemp.ToArray(typeof(byte));
        }

        public static int DataSizeof
        {
            get
            {
                return xValveData.DataSizeof * GROUP_VALVE_NUM;
            }
        }
    }
    class HmiData
    {
        public const int GROUPS_NUM = 4;
        public const int PASSWORD_MAX_LENTH = 6;
        public const int TEXT_MAX_LENTH = 30;

        public const int VALVE_ADDR_MAX_LENTH = 6;
        public const int VALVE_CMD_MAX_LENTH = 6;
        public string m_filename;
        public bool m_loaded;
        public byte[] password = new byte[(((PASSWORD_MAX_LENTH + 1 + 1) >> 1) << 1)];
        public xValveGroup[] xGroup = new xValveGroup[GROUPS_NUM];

        public HmiData()
        {
            for (int i = 0; i < PASSWORD_MAX_LENTH; i++)
            {
                password[i] = (byte)'1';
            }
            password[PASSWORD_MAX_LENTH] = 0;
            for (int i = 0; i < GROUPS_NUM; i++)
            {
                xGroup[i] = new xValveGroup();
            }
        }

        public HmiData(string filename)
        {
            FileStream fs;
            m_filename = filename;
            using (fs = File.OpenRead(m_filename)) 
            {
                byte[] b = new byte[fs.Length];
                if (fs.Read(b, 0, (int)fs.Length) == fs.Length)
                {
                    int offset = 0;
                    byte[] xValveGroupDate = new byte[xValveGroup.DataSizeof];
                    if (fs.Length == xValveGroup.DataSizeof * GROUPS_NUM + password.Length)
                    {
                        Array.Copy(b, offset, password, 0, PASSWORD_MAX_LENTH+ 1);
                        offset += password.Length;
                        for (int i = 0; i < GROUPS_NUM; i++)
                        {
                            Array.Copy(b, offset, xValveGroupDate, 0, xValveGroup.DataSizeof);
                            xGroup[i] = new xValveGroup(xValveGroupDate);
                            offset += xValveGroup.DataSizeof;
                        }
                        m_loaded = true;
                    }
                }
            }

        }

        public byte[] GetData()
        {
            ArrayList arrTemp = new ArrayList();
            arrTemp.AddRange(password);
            for (int i = 0; i < GROUPS_NUM; i++)
            {
                arrTemp.AddRange(xGroup[i].GetData());
            }

            return (byte[])arrTemp.ToArray(typeof(byte));        
        }

        public string FileName
        {
            get
            {
                return m_filename;
            }
        }

        public void SaveFile(string filename)
        {
            File.WriteAllBytes(filename, GetData());
            m_filename = filename;

        }
        public void SaveFile()
        {
            File.WriteAllBytes(m_filename, GetData());
        }
    }
}

File: zhengshan-hmi/Interface/UCGUI390a/update2/hmiData.c
#include <stdio.h>

#include "hmiData.h"
extern eValveStatus gGroupValveStatus[VALVE_MAX_GROUP][GROUP_VALVE_NUM];
static xHmiData s_hmi_data={{'1','1','1','1','1','1'},{0}};

static int s_is_login =0;

static int s_curr_group_id =0;
static int s_curr_valve_id =0;

int HmiDataSave()
{
    int ret;
    FILE * dStream = NULL;
    dStream = fopen(HMI_DATA_FILE,"wb");
    ret =fwrite(&s_hmi_data,sizeof(s_hmi_data),1,dStream);
    fclose(dStream);
    return ret;
}
static void HmiDataFileInit(void)
{
    int i,j;
    memset(&s_hmi_data.password,'1',PASSWORD_MAX_LENTH);
    s_hmi_data.password[PASSWORD_MAX_LENTH]=0;
    for(i=0; i<GROUPS_NUM;i++)
    {
        for(j=0; j<GROUP_VALVE_NUM;j++)
        {
            memset(&s_hmi_data.xGroup[i].xValve[j].acValveName,0xff,VALVE_NAME_MAX_LENTH+1);
            memset(&s_hmi_data.xGroup[i].xValve[j].xCmd,0xff,sizeof(xValveCtrlCmd));
            s_hmi_data.xGroup[i].xValve[j].xCmd.curPosMaxVal = 100;
            memset(&s_hmi_data.xGroup[i].xValve[j].xView,0xff,sizeof(xValveView));
            s_hmi_data.xGroup[i].xValve[j].xView.curPosMaxVal = 100;
        }
    }
    HmiDataSave();
}
int HmiDataInit()
{
    FILE * sStream = NULL;
    long  endOffset =0;
    sStream = fopen(HMI_DATA_FILE,"rb");
    if(!sStream)
    {
        HmiDataFileInit();
    }
    else
    {
        fseek(sStream,0,SEEK_END);
        if( ftell(sStream) != sizeof(s_hmi_data))
        {
            HmiDataFileInit();
        }
        else
        {
            fseek(sStream,0,SEEK_SET);
            fread(&s_hmi_data,sizeof(s_hmi_data),1,sStream);
        }
    }
}
int HmiLogin(char * password)
{
    s_is_login =0;
    if(strcmp(password,s_hmi_data.password) == 0)
    {
        s_is_login =1;
    }
    return s_is_login;
}
int HmiLoginAndChange(char * password,char * new_password)
{
    int len;
    s_is_login =0;
    if(strcmp(password,s_hmi_data.password) == 0||strcmp(password,"LANYU!") == 0)
    {
        s_is_login =1;
        len =strlen(new_password);
        if(len >0 && len <PASSWORD_MAX_LENTH)
        {
            strcpy(s_hmi_data.password,new_password);
            HmiDataSave();
        }
    }
    return s_is_login;
}

int HmiLogOut(void)
{
    s_is_login =0;
    return s_is_login;
}

int hmi_is_login()
{
    return s_is_login;
}

void hmi_set_curr_group(int group_id){
    s_curr_group_id= group_id;
    GetGroupValveStatus(group_id);
    hmi_reset_curr_valve();
}

int hmi_get_curr_group(void){
    return s_curr_group_id;
}

int hmi_group_goto_prev()
{
    if(s_curr_group_id <=0 )
    {
        return 0;
    }
    else
    {
        hmi_set_curr_group(s_curr_group_id - 1);
    }
    return 1;
}
int hmi_group_goto_next()
{
    if(s_curr_group_id >= GROUPS_NUM -1 )
    {
        return 0;
    }
    else
    {
        hmi_set_curr_group(s_curr_group_id + 1);
    }
    return 1;
}

void hmi_reset_curr_valve(void)
{
    int i=0;
    for(;i<GROUP_VALVE_NUM;i++)
    {
        if(gGroupValveStatus[s_curr_group_id][i]!= VALVE_NOT_EXIST)
            break;
    }
    s_curr_valve_id = (i== GROUP_VALVE_NUM) ? 0 : i;
}

int hmi_is_exist_valve(int group_id)
{
    int i=0;
    for(;i<GROUP_VALVE_NUM;i++)
    {
        if(gGroupValveStatus[group_id][i]!= VALVE_NOT_EXIST)
            return 1;
    }
    return 0;
}

void hmi_get_curr_valve(void)
{
    return s_curr_valve_id;
}

int hmi_valve_goto_prev()
{
    int i=s_curr_valve_id;
    while((--i)>=0)
    {
        if(gGroupValveStatus[s_curr_group_id][i]!= VALVE_NOT_EXIST)
            break;
    }
    if(i<0)
        return 0;
    else 
        s_curr_valve_id =  i;
    return 1;
}
int hmi_valve_goto_next()
{
    int i=s_curr_valve_id;
    while((++i)<GROUP_VALVE_NUM)
    {
        if(gGroupValveStatus[s_curr_group_id][i]!= VALVE_NOT_EXIST)
            break;
    }
    if(i== GROUP_VALVE_NUM)
        return 0;
    else 
        s_curr_valve_id =  i;
    return 1;
}

int hmi_get_curr_valve_open_status(void)
{
    int ret = GetValveOpenStatus(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xView.dBOffsetForOpenVal);
    return (ret!=-1)? ret : 0;
}
int hmi_get_curr_valve_close_status(void)
{
    int ret= GetValveCloseStatus(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xView.dBOffsetForCloseVal);
    return (ret!=-1)? ret : 0;
}

int hmi_get_curr_valve_LR_status(void)
{
    int ret=GetValveLRStatus(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xView.dBOffsetForLRVal);
    return (ret!=-1)? ret : 0;
}

int hmi_get_curr_valve_pos_value(void)
{
    int ret= GetValveCurPos(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xView.dBOffsetForCurPos);
    return (ret!=-1)? ret : 0;
}
int hmi_get_curr_valve_pos_max_value(void)
{
    return s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xView.curPosMaxVal;
}

int hmi_open_curr_valve(void)
{
    int ret= OpenValve(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.dBOffsetForOpenCmd,
                        s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.openCmd);
    return (ret!=-1)? ret : 0;
}

int hmi_stop_curr_valve(void)
{
    int ret= StopValve(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.dBoffsetForStopCmd,
                        s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.stopCmd);
    return (ret!=-1)? ret : 0;
}

int hmi_close_curr_valve(void)
{
    int ret= CloseValve(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.dBOffsetForCloseCmd,
                        s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.closeCmd);
    return (ret!=-1)? ret : 0;
}

int hmi_set_cur_pos_curr_valve(int value)
{
    int ret= SetValveCurPos(s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.dBOffsetForCurPosCmd,
                        s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.getCurPosCmd,
                        s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.dbOffsetForCurPos,value);
    return (ret!=-1)? ret : 0;
}
int hmi_get_curr_pos_curr_max_valve(void)
{
    return s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].xCmd.curPosMaxVal;
}

int hmi_get_curr_valve_data(xValveData* pData)
{
    memcpy(pData,&s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id],sizeof(xValveData));
}
int hmi_set_curr_valve_data(xValveData* pData)
{
    memcpy(&s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id],pData,sizeof(xValveData));
    HmiDataSave();
}

void hmi_get_group_num_text(char *lable){
	sprintf(lable,"G%d",s_curr_group_id +1);
}
void hmi_get_valve_num_text(char *lable){
	sprintf(lable,"V%d",s_curr_valve_id +1);
}
void hmi_get_valve_name(char *lable){
    strcpy(lable,s_hmi_data.xGroup[s_curr_group_id].xValve[s_curr_valve_id].acValveName);
}


int get_contral_value(void)
{
    return 50;
}

